lib/date_component.rb

```diff
@@ -1,44 +1,85 @@
+require 'date'
+
 # Component that parses date components from different fields into combined date range representations
 class DateComponent
-    attr_reader :date_str
-
-    def initialize
-        @start_year = nil
-        @end_year = nil
-        @start_month = nil
-        @end_month = nil
-        @start_day = nil
-        @end_day = nil
-
-        @date_str = ''
-    end
+    attr_reader :date_strs
 
-    def set_field(component, value)
-        return if /^0\-?$/.match value
+    # These are ISO-8601 season codes that are used in catalog records
+    @@season_codes = {
+        21 => 'Spring',
+        22 => 'Summer',
+        23 => 'Fall',
+        24 => 'Winter'
+    }
 
-        value_arr = value.split('-')
-        instance_variable_set("@start_#{component}", value_arr[0])
-        instance_variable_set("@end_#{component}", value_arr[1] || value_arr[0])
+    def initialize(date_values)
+        @date_values = date_values
+        @date_strs = {
+            start: nil,
+            end: nil
+        }
     end
 
     def create_strs
-        start_str = _format_str 'start'
-        end_str = _format_str 'end'
+        $logger.debug 'Parsing date strings from values', { values: @date_values }
+        # Split date strings into start/end values and then pivot them into properly arranged arrays
+        # e.g. [[1, 2], [3, 4], [5, 6]] to [[1, 3, 5], [2, 4, 6]]
+        date_components = @date_values.map { |v| _extract_date_components v }.transpose
+
+        begin
+            start_str = _transform_date_components_to_str date_components[0]
+            end_str = _transform_date_components_to_str date_components[1]
+            $logger.info "Setting date values: start / #{start_str}, end / #{end_str}"
 
-        if start_str == end_str
-            { start: start_str, end: nil }
-        else
-            { start: start_str, end: end_str }
+            @date_strs[:start] = start_str
+            @date_strs[:end] = end_str && end_str != start_str ? end_str : nil
+        rescue Date::Error, TypeError
+            $logger.error 'Unable to parse date values for this check-in box'
+            $logger.debug date_components
         end
     end
 
     private
 
-    def _format_str(pos)
-        year = instance_variable_get("@#{pos}_year")
-        month = instance_variable_get("@#{pos}_month") || '-'
-        day = instance_variable_get("@#{pos}_day")
+    def _extract_date_components(component)
+        component_arr = component.split('-')
+        component_arr[1] ? component_arr : [component_arr[0], component_arr[0]]
```

Comment:
Do you really mean to have `component_arr[0]` twice?

Reply:
Yeah, because of how these components interact, which might be clearer after adding a comment above, but the most common case is a set of values that look like this `['2020', '06-08', '01-30']` In that case we want `2020` to be applied to both dates, so we duplicate it and get this array: `[['2020', '2020'], ['06', '08'], ['01', '30']]`, which is then transposed so that we get `[['2020', '06', '01'], ['2020', '08', '30']]`